Docs: Revise IAM policies with more information (#989)
redpanda-data/documentation: modules/manage/pages/iceberg/access-iceberg-topics.adoc

:description: Learn how to access Redpanda topic data stored in Iceberg tables, using table metadata or a catalog integration.
:page-context-links: [{"name": "Linux", "to": "manage:iceberg/access-iceberg-topics.adoc" } ]
:page-categories: Iceberg, Tiered Storage, Management, High Availability, Data Replication, Integration

To read from the Redpanda-generated xref:manage:iceberg/topic-iceberg-integration.adoc[Iceberg table], your Iceberg-compatible client or tool needs access to the catalog to know the current state of the table. The catalog provides the location of the current metadata pointer. You can configure Redpanda to either store the table metadata in https://iceberg.apache.org/javadoc/1.5.0/org/apache/iceberg/hadoop/HadoopCatalog.html[HadoopCatalog^] format in the same object storage bucket or container, or connect to a REST-based catalog. 

Set the cluster configuration property `iceberg_catalog_type` with one of the following values:

* `object_storage` (default): Write catalog files to the same object storage bucket as the data files. Use the object storage URL with an Iceberg client to access the catalog and data files for your Redpanda Iceberg tables.

> * `rest`: Connect to and update an Iceberg catalog using a REST API. See the https://github.com/apache/iceberg/blob/main/open-api/rest-catalog-open-api.yaml[Iceberg REST Catalog API specification].

Reviewer: ```suggestion
* `rest`: Connect to and update an Iceberg catalog using a REST API. See the https://github.com/apache/iceberg/blob/main/open-api/rest-catalog-open-api.yaml[Iceberg REST Catalog API specification^].
```

> The quickest way to access the data in Iceberg-enabled topics is to use the `object_storage` catalog type, and point your query engine to the table metadata path in the bucket or container. For production use cases, it is recommended to use a REST catalog integration to more easily manage governance, and ensure discoverability and consistency across multiple engines or tools. In either case, you use the catalog to load, query, or refresh the Iceberg data as you produce to the Redpanda topic. For specific guidance on adding the Iceberg tables to your data warehouse or lakehouse using the catalog, refer to the official documentation of your query engine or Iceberg-compatible tool. 

Reviewer: ```suggestion
The quickest way to access the data in Iceberg-enabled topics is to use the `object_storage` catalog type, and point your query engine to the table metadata path in the bucket or container. For production use cases, it is recommended to use a REST catalog integration to more easily manage governance, and ensure discoverability and consistency across multiple engines or tools. In either case, you use the catalog to load, query, or refresh the Iceberg data as you produce to the Redpanda topic. See the documentation for your query engine or Iceberg-compatible tool for specific guidance on adding the Iceberg tables to your data warehouse or lakehouse using the catalog. 
```

> Once you have xref:manage:iceberg/topic-iceberg-integration.adoc#enable-iceberg-integration[enabled the Iceberg integration] for a topic and selected a catalog type, you cannot switch to another catalog type.

Reviewer: ```suggestion
After you have xref:manage:iceberg/topic-iceberg-integration.adoc#enable-iceberg-integration[enabled the Iceberg integration] for a topic and selected a catalog type, you cannot switch to another catalog type.
```

> = Access Topic Data in Iceberg Tables

Reviewer: @kbatuigas the Access page the Query page are missing the admonition at the top of the page about needing an Enterprise license.

Reviewer 2: +1

Reviewer 3: Updated

> = Access Topic Data in Iceberg Tables

Reviewer 2: I think this page needs to be entitled 'Working with Iceberg catalogs' as it explains the catalog types, functions and also how to configure catalog integration from the Redpanda side. This isn't really the 'accessing' part on this page (for the most part).  "accessing" is really the querying IMO

Reviewer 2: Overall I like the new 3 page structure though!

Reviewer 3: Updated to "Use Iceberg Catalogs" (chose "Use" to follow our style guidelines for titles and headings

:description: Learn how to access Redpanda topic data stored in Iceberg tables, using table metadata or a catalog integration.
:page-categories: Iceberg, Tiered Storage, Management, High Availability, Data Replication, Integration

> To read from the Redpanda-generated xref:manage:iceberg/topic-iceberg-integration.adoc[Iceberg table], your Iceberg-compatible client or tool needs access to the catalog to retrieve the table metadata and know the current state of the table. The catalog provides the location of the current metadata pointer. You can configure Redpanda to either connect to a REST-based catalog, or use a file-system based catalog. 

Reviewer 4: >The catalog provides the location of the current metadata pointer.

The catalog provides the current table metadata, which includes locations for all of the table's data files.

> The quickest way to access the data in Iceberg-enabled topics is to use the file-system based catalog, and point your query engine to the table metadata path in the bucket or container. However, for production use cases, Redpanda recommends a REST catalog integration to more easily manage governance, and ensure discoverability and consistency across multiple engines or tools. In either case, you use the catalog to load, query, or refresh the Iceberg table as you produce to the Redpanda topic. See the documentation for your query engine or Iceberg-compatible tool for specific guidance on adding the Iceberg tables to your data warehouse or lakehouse using the catalog. 

Reviewer 4: Maybe can we flip the paragraph structure to roughly:

- For production usage, Redpanda recommends using a REST catalog to manage metadata (better consistency guarantees, governance, etc)
- If this isn't possible you may use the file-system based catalog which does not require maintaining a separate service.

I think the current structure leads someone skimming into believing the filesystem catalog is better than it is

Reviewer 3: Good callout; updated

> For production usage, Redpanda recommends a REST catalog integration to manage Iceberg metadata so you can more easily facilitate governance, ensure discoverability, and maintain consistency across multiple engines or tools. However, if it is not possible to use a REST catalog, you may use the file-system based catalog, which does not require you to maintain a separate service to access the Iceberg data. In either case, you use the catalog to load, query, or refresh the Iceberg table as you produce to the Redpanda topic. See the documentation for your query engine or Iceberg-compatible tool for specific guidance on adding the Iceberg tables to your data warehouse or lakehouse using the catalog. 

Reviewer 2: "For production deployments, Redpanda recommends using an external REST catalog to manage Iceberg metadata, which maximizes discoverability, facilitates data governance, allows you take advantage of your catalog's built-in table maintenance features, and safely handles concurrent access to tables from multiple engines and/or tools that support Iceberg."

Reviewer 3: Updated

> After you have xref:manage:iceberg/topic-iceberg-integration.adoc#enable-iceberg-integration[enabled the Iceberg integration] for a topic and selected a catalog type, you cannot switch to another catalog type.

Reviewer 2: This reads a little confusing wrt the order of operations: you first enable iceberg at the cluster level , select the catalog type at the cluster level (with  cluster configs) and THEN enable iceberg for a topic  (not the other way around as stated here).

Reviewer 3: Updated

> == Integrate REST catalog

Reviewer 2: Perhaps just 'Connect to a REST catalog'

Reviewer 3: Updated

> Redpanda supports connecting to and updating an Iceberg catalog using a REST API, compliant with the https://github.com/apache/iceberg/blob/main/open-api/rest-catalog-open-api.yaml[Iceberg REST Catalog API specification^]. Use this catalog integration type with REST-enabled Iceberg catalog services, such as https://docs.databricks.com/en/data-governance/unity-catalog/index.html[Databricks Unity^] and https://other-docs.snowflake.com/en/opencatalog/overview[Snowflake Open Catalog^].

Reviewer 2: "Redpanda supports connecting to an Iceberg REST Catalog, using the standard REST API supported by many catalog providers.  Use this catalog .....

Reviewer 3: Updated

> To integrate a REST catalog, set the following cluster configuration properties:

Reviewer 2: 'To configure a connection to a REST catalog, set the following' ....

Reviewer 3: Updated

* config_ref:iceberg_catalog_type,true,properties/cluster-properties[`iceberg_catalog_type`]: `rest`
* config_ref:iceberg_rest_catalog_endpoint,true,properties/cluster-properties[`iceberg_rest_catalog_endpoint`]: The endpoint URL for your Iceberg catalog, which you either manage directly, or is managed by an external catalog service.
* config_ref:iceberg_rest_catalog_client_id,true,properties/cluster-properties[`iceberg_rest_catalog_client_id`]: The ID to connect to the REST server.

> * config_ref:iceberg_rest_catalog_client_secret,true,properties/cluster-properties[`iceberg_rest_catalog_client_secret`]: The secret data to connect to the REST server.

Reviewer 4: nit: any reason to use "REST server" instead of "REST catalog"?

For REST catalogs that use self-signed certificates, also configure these properties:

> * config_ref:iceberg_rest_catalog_trust_file,true,properties/cluster-properties[`iceberg_rest_catalog_trust_file`]: The path to a file containing a certificate chain to trust for the REST catalog.

Reviewer 2: @deniscoady @michael-redpanda  double checking here, trust file is only needed for self-signed certs?  Or rather any time you need to configure what Redpanda should trust?

----
spark.sql.catalog.streaming.type = hadoop
spark.sql.catalog.streaming.warehouse = s3a://<bucket-name>/path/to/redpanda-iceberg-table
----

Depending on your processing engine, you may need to specify the location for the metadata to xref:manage:iceberg/query-iceberg-tables.adoc[query the Iceberg table]. You may also need to create a new table in your data warehouse or lakehouse for the Iceberg data.

=== Specify metadata location

The config_ref:iceberg_catalog_base_location,true,properties/cluster-properties[] property stores the base path for the file-system based catalog if using the `object_storage` catalog type. The default value is `redpanda-iceberg-catalog`. 

CAUTION: Do not change the `iceberg_catalog_base_location` value after you have enabled Iceberg integration for a topic.

If your engine needs the full JSON metadata path, use the following:

```
redpanda-iceberg-catalog/metadata/redpanda/<topic-name>/v<version-number>.metadata.json
```

This provides read access to a single snapshot (denoted by `version-number`) of a table. 

NOTE: Redpanda automatically removes expired snapshots on a periodic basis. 

> == REST catalog integration

Reviewer 5: Nit: I'd move this higher on this page. Document first the production grade solution, then the "lesser" ones so the user doesn't get started on the "wrong" path.

> spark.sql.catalog.streaming.uri = http://catalog-service:8181

Reviewer 2: @kbatuigas this section is great. I'm wondering if you truly don't need to provide credentials here from the Spark side (you've set them on the Redpanda side).  Can you double-check?

Reviewer 3: @mattschumpert I can't seem to find official documentation for it, but in addition to setting some Spark config properties it seems that the best practice is to just use env vars. I can add a note

Reviewer 3: Added in https://github.com/redpanda-data/docs/pull/989/commits/e8e54b3271d1071f6d6384cc10612a7333850313. Could use SME confirmation

> For production use cases, Redpanda recommends the `rest` option with REST-enabled Iceberg catalog services such as https://docs.tabular.io/[Tabular^], https://docs.databricks.com/en/data-governance/unity-catalog/index.html[Databricks Unity^] and https://other-docs.snowflake.com/en/opencatalog/overview[Snowflake Open Catalog^].

Reviewer 5: Tabular is being sunsetted afaik.

> Using Spark SQL, you can query the Iceberg table directly by specifying the catalog name:

Reviewer 2: wouldn't it just be 'streaming.<topic-name>' ?  (considering the catalog is just called 'streaming'?  Where does 'redpanda' figure in this example?

Reviewer 2: NM, I see it below.   Maybe a better way to word this then is 

"Using Spark SQL, you can query the Iceberg table directly by specifying the catalog name, followed by the namespace and table name.  Redpanda pouts the iceberg table into a namespace called 'redpanda''

Reviewer 3: Updated

[,sql]
----

> SELECT * FROM streaming.redpanda.<topic-name>;

Reviewer 4: Maybe it's worth explicitly calling out that Redpanda will put the Iceberg topic's table into a namespace called "redpanda", creating the namespace if necessary

(this isn't currently configurable)

----

Spark can use the REST catalog to automatically discover the topic's Iceberg table. 

> Depending on your processing engine, to xref:manage:iceberg/query-iceberg-topics.adoc[query the Iceberg data] you may need to also create a new table in your data warehouse or lakehouse. For an example, see xref:manage:iceberg/redpanda-topics-iceberg-snowflake-catalog.adoc[].

Reviewer 4: >also create a new table in your data warehouse or lakehouse

I know this is technical true, but it kind of reads like Redpanda isn't creating a table anywhere. Maybe can we change to something to the effect of

Depending on the processing engine and your Iceberg catalog implementation, you may need to create a table in the query engine to point your data lakehouse at the table in the catalog.

Reviewer 3: Updated

> Depending on your processing engine, you may need to also create a table in the engine to point the data lakehouse at the table in the catalog. For an example, see xref:manage:iceberg/redpanda-topics-iceberg-snowflake-catalog.adoc[].

Reviewer: ```suggestion
Depending on your processing engine, you may need to also create a table in the engine to point the data lakehouse at the table in the catalog. For an example, see xref:manage:iceberg/redpanda-topics-iceberg-snowflake-catalog.adoc[].
```
This isn't clear to me. Is it missing a word?

Reviewer 3: Updated

> Depending on your processing engine, you may need to also create a table in the engine to point the data lakehouse at the table in the catalog. For an example, see xref:manage:iceberg/redpanda-topics-iceberg-snowflake-catalog.adoc[].

Reviewer 2: Text looks good.  But didn't you say you eventually  got Snowflake to work without creating the table explicitly ?

Reviewer 3: @mattschumpert no, Snowflake required creating the external table https://docs.redpanda.com/current/manage/iceberg/redpanda-topics-iceberg-snowflake-catalog/#create-iceberg-table-in-snowflake I think Spark and ClickHouse didn't require it.

== Integrate file-system based catalog (`object_storage`)

By default, Iceberg topics use the file-system based catalog (config_ref:iceberg_catalog_type,true,properties/cluster-properties[`iceberg_catalog_type`] cluster configuration set to `object_storage`). Redpanda stores the table metadata in https://iceberg.apache.org/javadoc/1.5.0/org/apache/iceberg/hadoop/HadoopCatalog.html[HadoopCatalog^] format in the same object storage bucket or container as the data files.

> If using the `object_storage` catalog type, you provide the object storage URL to an Iceberg client so it can access the catalog and data files for your Redpanda Iceberg tables.

Reviewer 2: 'you provide the object storage URI of a table's metadata.json file to an Iceberg client ....'

Reviewer 3: updated

> === Specify metadata location

Reviewer 2: This is kind of an advanced (and certainly optional) configuration that should probably go at the end of this section.  In most cases we'd rather not have the user be customizing this, but only if its necessary in order to work with a particular iceberg client.

Reviewer 3: Moved to end of section

The config_ref:iceberg_catalog_base_location,true,properties/cluster-properties[`iceberg_catalog_base_location`] property stores the base path for the file-system based catalog if using the `object_storage` catalog type. The default value is `redpanda-iceberg-catalog`. 

CAUTION: Do not change the `iceberg_catalog_base_location` value after you have enabled Iceberg integration for a topic.

If your engine needs the full JSON metadata path, use the following:

```
redpanda-iceberg-catalog/metadata/redpanda/<topic-name>/v<version-number>.metadata.json
```

> This provides read access to a single snapshot (denoted by `version-number`) of a table. 

Reviewer 4: This isn't quite the right terminology because a single instance of the table metadata will contain many snapshots of the table.

How about:

This provides read access all snapshots in the the given version of the table.

> NOTE: Redpanda automatically removes expired snapshots on a periodic basis. 

Reviewer 4: nit: this maybe a better fit for the Query Iceberg Topics section

Reviewer 3: Added link to a new time travel section in [Query Iceberg Topics](https://deploy-preview-989--redpanda-docs-preview.netlify.app/25.1/manage/iceberg/query-iceberg-topics/#query-using-time-travel) which also mentions snapshot expiry

> This provides read access to all snapshots in the given version (denoted by `version-number`) of a table. 

Reviewer 2: 'this provides read access' isn't really true.  

"The version refers to a specific snapshot. Unfortunately, most Iceberg clients using the 'object_storage' catalog type must read Table data from a specific snapshot, requiring the client to be updated each time new data is committed to a table

Reviewer 4: >The version refers to a specific snapshot.

This isn't quite true. Each new version of the table refers to a set of snapshots that may be queried by query engines for that table. Query engines need only a single metadata JSON to access the many historical snapshots available for time travel queries.

Maybe it would capture what you're trying to say as:

"This provides read access to all snapshots written as of the given version (denoted by `version-number`) of a table."

Reviewer 3: Updated

=== Example file-system based catalog configuration

To configure Apache Spark to use a file system-based catalog, specify at least the following properties:

[,spark]
----
spark.sql.catalog.streaming = org.apache.iceberg.spark.SparkCatalog
spark.sql.catalog.streaming.type = hadoop
# AWS S3 bucket used for object storage
spark.sql.catalog.streaming.warehouse = s3a://<bucket-name>/redpanda-iceberg-catalog
# You may need to configure additional properties based on your object storage provider.
# See https://spark.apache.org/docs/latest/configuration.html
----

> Depending on your processing engine, you may need to also create a new table to point the data lakehouse at the table location.

Reviewer: ```suggestion
Depending on your processing engine, you may need to also create a new table to point the data lakehouse at the table location.
```
Should this be ...point the data lakehouse to the table location?

Reviewer 3: Updated

[,yaml]
----
iceberg_catalog_type: rest 
iceberg_rest_catalog_endpoint: http://catalog-service:8181 
iceberg_rest_catalog_client_id: <rest-connection-user>
iceberg_rest_catalog_client_secret: <rest-connection-password>
----

And you use Apache Spark as a processing engine, configured to use a catalog named `streaming`:

[,spark]
----
spark.sql.catalog.streaming = org.apache.iceberg.spark.SparkCatalog
spark.sql.catalog.streaming.type = rest
spark.sql.catalog.streaming.uri = http://catalog-service:8181
----

Using Spark SQL, you can query the Iceberg table directly by specifying the catalog name:

[,sql]
----
SELECT * FROM streaming.redpanda.<topic-name>;
----

Spark can use the REST catalog to automatically discover the topic's Iceberg table. Redpanda puts the Iceberg table into a namespace called `redpanda`, creating the namespace if necessary.

Depending on your processing engine, you may need to also create a table in the engine to point the data lakehouse to the table location in the catalog. For an example, see xref:manage:iceberg/redpanda-topics-iceberg-snowflake-catalog.adoc[].

> == Integrate file-system based catalog (`object_storage`)

Reviewer 2: 'Configure a file-system based catalog'

Reviewer 3: Updated

By default, Iceberg topics use the file-system based catalog (config_ref:iceberg_catalog_type,true,properties/cluster-properties[`iceberg_catalog_type`] cluster configuration set to `object_storage`). Redpanda stores the table metadata in https://iceberg.apache.org/javadoc/1.5.0/org/apache/iceberg/hadoop/HadoopCatalog.html[HadoopCatalog^] format in the same object storage bucket or container as the data files.

If using the `object_storage` catalog type, you provide the object storage URL to an Iceberg client so it can access the catalog and data files for your Redpanda Iceberg tables.

=== Specify metadata location

The config_ref:iceberg_catalog_base_location,true,properties/cluster-properties[`iceberg_catalog_base_location`] property stores the base path for the file-system based catalog if using the `object_storage` catalog type. The default value is `redpanda-iceberg-catalog`. 

CAUTION: Do not change the `iceberg_catalog_base_location` value after you have enabled Iceberg integration for a topic.

> If your engine needs the full JSON metadata path, use the following:

Reviewer 2: This seems out of place.  We explain the object store catalog and how to configure it (including the optional path). It's fine to explain here that the catalog generates a metadata.json for each snapshot but instructing the user about using this metadata path should be part of the 'querying iceberg topics' phase.

Reviewer 3: Moved to Querying Iceberg Topics doc

```
redpanda-iceberg-catalog/metadata/redpanda/<topic-name>/v<version-number>.metadata.json
```

This provides read access to all snapshots in the given version (denoted by `version-number`) of a table. 

> NOTE: Redpanda automatically removes expired snapshots on a periodic basis. Snapshot expiry helps maintain a smaller metadata size and reduces the window available for xref:manage:iceberg/query-iceberg-topics.adoc#query-using-time-travel[time travel].  

Reviewer 2: Is the last part true @andrwng ?  Are we narrowing the window for time travel queries or just expiring completely unused snapshot files that couldn't be used in a time travel query anyways?  Thats what I understood from our chat last week

Reviewer 4: What's written is correct. It is by default set to retaining snapshots for 5 days, meaning you can run time travel queries as of up to 5 days ago.

If we want to be explicit about the behavior, we will honor the following Iceberg table properties:
- history.expire.max-snapshot-age-ms
- history.expire.min-snapshots-to-keep
- history.expire.max-ref-age-ms

from https://apache.github.io/iceberg/docs/1.4.3/configuration/#table-behavior-properties

> just expiring completely unused snapshot files that couldn't be used in a time travel query anyways

This feature would be orphaned file cleanup, which we don't have yet.

Reviewer 3: @mattschumpert can you provide guidance on what to document here? I'd imagine that we'd say you can do time travel for up to 5 days ago by default

=== Example file-system based catalog configuration

To configure Apache Spark to use a file system-based catalog, specify at least the following properties:

[,spark]
----
spark.sql.catalog.streaming = org.apache.iceberg.spark.SparkCatalog
spark.sql.catalog.streaming.type = hadoop
# AWS S3 bucket used for object storage
spark.sql.catalog.streaming.warehouse = s3a://<bucket-name>/redpanda-iceberg-catalog
# You may need to configure additional properties based on your object storage provider.

> # See https://spark.apache.org/docs/latest/configuration.html

Reviewer 2: 👍

Once you have xref:manage:iceberg/topic-iceberg-integration.adoc#enable-iceberg-integration[enabled the Iceberg integration] for a topic and selected a catalog type, you cannot switch to another catalog type.

== REST catalog integration

> For production use cases, Redpanda recommends the `rest` option with REST-enabled Iceberg catalog services such as https://docs.databricks.com/en/data-governance/unity-catalog/index.html[Databricks Unity^] and https://other-docs.snowflake.com/en/opencatalog/overview[Snowflake Open Catalog^].

Reviewer: ```suggestion
For production use cases, Redpanda recommends the `rest` option with REST-enabled Iceberg catalog services, such as https://docs.databricks.com/en/data-governance/unity-catalog/index.html[Databricks Unity^] and https://other-docs.snowflake.com/en/opencatalog/overview[Snowflake Open Catalog^].
```

Depending on your processing engine, to xref:manage:iceberg/query-iceberg-tables.adoc[query the Iceberg data] you may need to create a new table in your data warehouse or lakehouse. See also: xref:manage:iceberg/redpanda-topics-iceberg-snowflake-catalog.adoc[]

> == File system-based catalog (`object_storage`)

Reviewer: Suggest renaming these headings to **Integrate with file system catalog** and **Integrate with REST catalog**.

If you are using the `object_storage` catalog type, you must also set up the catalog integration in your processing engine using the location of the metadata in the bucket or container.

> For example, you can configure Spark to use a file system-based catalog with at least the following properties, is using AWS S3 for object storage:

Reviewer: This sentence doesn't make sense.

Reviewer 3: Updated

> .Example Spark configuration for `object_storage` catalog type

Reviewer: I'd either remove this or rephrase the prior sentence. If we're leading into an example with "For example... :" I don't we also need a title for the example.
